app/src/actioncommands.cpp

```diff
@@ -563,6 +566,215 @@ void ActionCommands::duplicateKey()
+        startL = mEditor->playback()->getRangedStartFrame();
+        stopL = mEditor->playback()->getRangedEndFrame();
+    }
+    LayerManager* layerMgr = mEditor->layers();
+    int lIndex = layerMgr->currentLayer()->type();
+    //     BITMAP = 1     VECTOR = 2   //not yet      SOUND = 4     CAMERA = 5
+    if (!(lIndex == 1 || lIndex == 2)) //not yet || lIndex == 4 || lIndex == 5)
+    {
+        QMessageBox msgBox;
+        msgBox.setText(tr("Can only be used on Bitmap and Vector layers."));
+        msgBox.exec();
+        return;
+    }
+    CopyMultiplekeyframesDialog* cd = new CopyMultiplekeyframesDialog(layerMgr, startL, stopL, new QWidget);
+    cd->exec();
+
+    // If validation is NOT OK -> return
+    if (cd->result() != QDialog::Accepted) { return; }
+    if (!cd->getValidity())
+    {
+        QMessageBox msgBox;
+        msgBox.setWindowTitle(tr("Action not valid!"));
+        msgBox.setText(tr("Failed validation check:\n- Exceeds 9999 Frames OR\n- Range not valid"));
+        msgBox.exec();
+        return;
+    }
+
+    // If Validation is OK
+    startL = cd->getFirstFrame();
+    stopL = cd->getLastFrame();
+    Layer *fromLayer = layerMgr->findLayerByName(cd->getFromLayer());
+    Q_ASSERT(fromLayer != nullptr);
+    Layer *toLayer = nullptr;
+    int scrubOrg = mEditor->currentFrame();
+    int num, startF;
+    switch (cd->getCurrentTab()) {
+    case 0: // Copy
+        toLayer = layerMgr->findLayerByName(cd->getCopyToLayer());
+        Q_ASSERT(toLayer != nullptr);
+        num = cd->getNumLoops();
+        startF = cd->getCopyStartFrame();
+        copyFrames(startL, stopL, num, startF, fromLayer, toLayer);
+        break;
+    case 1: // Move
+        startF = cd->getMoveStartFrame();
+        toLayer = layerMgr->findLayerByName(cd->getMoveToLayer());
+        Q_ASSERT(toLayer != nullptr);
+        moveFrames(startL, stopL,startF, fromLayer, toLayer);
+        break;
+    case 2: // Reverse
+        startF = cd->getReverseStartFrame();
+        toLayer = layerMgr->findLayerByName(cd->getFromLayer());
+        Q_ASSERT(toLayer != nullptr);
+        reverseFrames(startL, stopL, startF, toLayer);
+        break;
+    case 3: // Delete
+        deleteFrames(startL, stopL, fromLayer);
+        break;
+    default:
+        Q_ASSERT(false);
+        break;
+    }
+    mEditor->scrubTo(scrubOrg);
+    mEditor->layers()->notifyLayerChanged(toLayer);
+    mEditor->layers()->notifyAnimationLengthChanged();
+}
+
+void ActionCommands::copyFrames(int startL, int stopL, int loops, int startAt, Layer *fLayer, Layer *tLayer)
+{
+    int num = loops;
+    int startF = startAt;
+    for (int i = 0; i < num; i++)
+    {
+        for (int j = startL; j < stopL + 1; j++, startF++)
+        {
+            mEditor->scrubTo(j);
+            if (fLayer->keyExists(j))
+            {
+                KeyFrame* kf = fLayer->getKeyFrameAt(j);
+                if (kf == nullptr) return;
```

Comment:
Since key 'j' exists, there should not be a case where it became a nullptr, but it's much like the Q_ASSERT-thing that I still don't understand fully.
I'd appreciate some input too.

Reply:
Considering that you have a keyExists above, I'd say you should be able to drop the  "if (kf == nullptr) return;"

Since you're just checking for nullptr, you could use Q_CHECK_PTR. if the app crashes, then it'll tell you at the exact line number in the console. Q_ASSERT will do the same except it's a boolean expression.

http://doc.qt.io/qt-5/debug.html#debugging-macros

The macros serves as a mean to help you debug the application, especially when it crashes with a useless message.